runtime/protocol/http-client/common/src/aws/smithy/kotlin/runtime/http/interceptors/RequestCompressionInterceptor.kt

```diff
@@ -0,0 +1,71 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package aws.smithy.kotlin.runtime.http.interceptors
+
+import aws.smithy.kotlin.runtime.InternalApi
+import aws.smithy.kotlin.runtime.client.ProtocolRequestInterceptorContext
+import aws.smithy.kotlin.runtime.compression.CompressionAlgorithm
+import aws.smithy.kotlin.runtime.compression.compressRequest
+import aws.smithy.kotlin.runtime.http.HttpBody
+import aws.smithy.kotlin.runtime.http.request.HttpRequest
+import aws.smithy.kotlin.runtime.telemetry.logging.logger
+import kotlin.coroutines.coroutineContext
+
+private val VALID_COMPRESSION_THRESHOLD_BYTES_RANGE = 0..10_485_760
+
+/**
+ * HTTP interceptor that compresses request payloads
+ * @param compressionThresholdBytes The threshold for applying compression to a request
+ * @param supportedCompressionAlgorithms The ID's of compression algorithms that are supported by the server
+ * @param availableCompressionAlgorithms The compression algorithms that are supported by the client
+ */
+@InternalApi
+public class RequestCompressionInterceptor(
+    private val compressionThresholdBytes: Long,
+    private val availableCompressionAlgorithms: List<CompressionAlgorithm>,
+    private val supportedCompressionAlgorithms: List<String>,
+) : HttpInterceptor {
+
+    init {
+        require(compressionThresholdBytes in VALID_COMPRESSION_THRESHOLD_BYTES_RANGE) { "compressionThresholdBytes ($compressionThresholdBytes) must be in the range $VALID_COMPRESSION_THRESHOLD_BYTES_RANGE" }
+    }
+
+    override suspend fun modifyBeforeRetryLoop(
+        context: ProtocolRequestInterceptorContext<Any, HttpRequest>,
+    ): HttpRequest {
+        val payloadSizeBytes = context.protocolRequest.body.contentLength
+        val algorithm = findMatch(supportedCompressionAlgorithms, availableCompressionAlgorithms)
+
+        return if (algorithm != null && (context.protocolRequest.body.isStreaming || payloadSizeBytes?.let { it >= compressionThresholdBytes } == true)) {
+            algorithm.compressRequest(context.protocolRequest)
+        } else {
+            val logger = coroutineContext.logger<RequestCompressionInterceptor>()
+            val skipCause = if (algorithm == null) "requested compression algorithm(s) are not supported by the client" else "request size threshold ($compressionThresholdBytes) was not met"
```

Comment:
correctness: "requested compression algorithm(s) are not supported by the **service**"

Reply:
wouldn't it be the client? because different clients can have different supported algorithms. Why the service?

Reply:
If I understand the wording correctly, the client is the one requesting compression algorithms to be used. A "requested compression algorithm" is always supported by the client.

I think it can be reworded to "no **modeled** compression algorithms are supported by the client" or "(client's) requested compression algorithms are not supported by the **service**"